### ISSUE TITLE
PG::NotNullViolation: ERROR: column "created_at" contains null values 

### ISSUE THREAD

**Issue by 99cm** [spree/spree#7857]
AddNullFalseToSpreeVariantsTimestamps: migrating - 
change_column_null(:spree_variants, :created_at, false)
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

PG::NotNullViolation: ERROR:  column "created_at" contains null values
: ALTER TABLE "spree_variants" ALTER "created_at" SET NOT NULL

## Context

original spree/core/db/migrate/20161014152814_add_null_false_to_spree_variants_timestamps.spree.rb
'class AddNullFalseToSpreeVariantsTimestamps < ActiveRecord::Migration[5.0]
  def change
    change_column_null :spree_variants, :created_at, false
    change_column_null :spree_variants, :updated_at, false
  end
end'

## Steps to Reproduce

1.
2.
3.
4.

## Your Environment

* Version used:
* Gemfile and Gemfile.lock as text in a Gist:
* Any relevant stack traces ("Full trace" preferred):

**Comment by nnande** [comment]
@99cm Could you provide more details about this issue, please?
It would be helpful if you filled **Steps to Reproduce** section especially

**Comment by 99cm** [comment]
I run:
rake db:migrate

when processing db/migrate/20161014152814_add_null_false_to_spree_variants_timestamps.spree.rb
Rake aborted and canceled later on migrations.
Rake migrations unable to completed.

**Comment by 99cm** [comment]
I searched online. It says 'change_column_null' should have 4 parameters.

**Comment by nnande** [comment]
@99cm Thanks for your feedback.

According to change_column_null method definition, it's true it defines 4 parameters, but note that the last one is optional. I do not think that it's the problem in our case.

In my opinion, the problem is that there are some Variant records present in your database with ``created_at`` or ``updated_at`` set to nil, which makes not null constraints creation fail.

You have to ensure that your database has no Variant records with created_at or updated_at values set to nil before running the migration.

If your database contains such, then you should try:
 - updating their timestamp columns with some values, let's say:

``Spree::Variant.where(created_at: nil).update(created_at: Time.zone.now)``
``Spree::Variant.where(updated_at: nil).update(updated_at: Time.zone.now)``

 - or deleting these variants before running the migration. 

Please let me know if it helped.

**Comment by 99cm** [comment]
@nnande 
> 
    updating their timestamp columns with some values, let's say:

Spree::Variant.where(created_at: nil).update(created_at: Time.zone.now)
Spree::Variant.where(updated_at: nil).update(updated_at: Time.zone.now)

It seems that these should be done by migrating 20161014145148_add_created_at_to_variant.spree.rb 

class AddCreatedAtToVariant < ActiveRecord::Migration[5.0]
  def change
    add_column :spree_variants, :created_at, :datetime
    Spree::Variant.reset_column_information
    Spree::Variant.where.not(updated_at: nil).update_all('created_at = updated_at')
    Spree::Variant.where(updated_at: nil).update_all(created_at: Time.current, updated_at: Time.current)
  end
end

**Comment by nnande** [comment]
@99cm In fact, the migration you mentioned above does affect only default-scoped ``Variants``.

``Spree::Variant`` class makes use of acts_as_paranoid gem, which attaches default scope to it, showing only "non-destroyed" records, conceptually something like: 

```
module Spree
  class Variant < Spree::Base
    default_scope { where(deleted_at: nil) }
```

It's very likely that your database consists of Variant record with deleted_at set to some not nil value, which is not affected by `20161014145148` migration , causing `20161014152814` migration failure.

There is a commit introduced on Feb 18 that fixes the `20161014145148` migration, so that it affects all the Variants present (unscoped): https://github.com/spree/spree/commit/9d39477eba385f0998927efaa26de12357217c9f

I would recommend you upgrading your Spree to the current version and try running migrations afterwards.

**Comment by 99cm** [comment]
@nnande solutions did not solve the issue. Actually "spree_variants" table keeps all deleted spree_variant samples that contain null values. I use phppgadmin to empty all deleted spree_variant samples. This solved the issue.
